pymc_experimental/sampling/optimizations/conjugacy.py

```diff
@@ -0,0 +1,153 @@
+
+        elif isinstance(client.op, DimShuffle) and client.op.is_left_expand_dims:
+            for sub_client, sub_inp_idx in clients[client.outputs[0]]:
+                if isinstance(sub_client.op, used_by_type) and sub_inp_idx in used_as_arg_idx:
+                    # RV is broadcasted and then used by the RV type
+                    used_by.append(sub_client.default_output())
+                elif strict:
+                    # Some other unrecognized use, bail out
+                    return []
+        elif strict:
+            # Some other unrecognized use, bail out
+            return []
+
+    return used_by
+
+
+def wrap_rv_and_conjugate_rv(
+    fgraph: FunctionGraph, rv: Variable, conjugate_rv: Variable, inputs: Sequence[Variable]
+) -> Variable:
+    """Wrap the RV and its conjugate posterior RV in a ConjugateRV node.
+
+    Also takes care of handling the random number generators used in the conjugate posterior.
+    """
+    rngs, next_rngs = zip(*collect_default_updates(conjugate_rv, inputs=[rv, *inputs]).items())
+    for rng in rngs:
+        if rng not in fgraph.inputs:
+            fgraph.add_input(rng)
+    conjugate_op = ConjugateRV(inputs=[rv, *inputs, *rngs], outputs=[rv, conjugate_rv, *next_rngs])
+    return conjugate_op(rv, *inputs, *rngs)[0]
+
+
+def create_untransformed_free_rv(
+    fgraph: FunctionGraph, rv: Variable, name: str, dims: Sequence[str | Variable]
+) -> Variable:
+    """Create a model FreeRV without transform."""
+    transform = None
+    value = rv.type(name=name)
+    fgraph.add_input(value)
+    free_rv = model_free_rv(rv, value, transform, *dims)
+    free_rv.name = name
+    return free_rv
+
+
+@node_rewriter(tracks=[ModelFreeRV])
+def beta_binomial_conjugacy(fgraph: FunctionGraph, node):
+    [beta_free_rv] = node.outputs
+    beta_rv, beta_value, *beta_dims = node.inputs
+
+    if not isinstance(beta_rv.owner.op, Beta):
+        return None
+
+    p_arg_idx = 3  # inputs to Binomial are (rng, size, n, p)
+    binomial_rvs = rv_used_by(fgraph, beta_free_rv, Binomial, p_arg_idx)
+
+    if len(binomial_rvs) != 1:
+        # Question: Can we apply conjugacy when RV is used by more than one binomial?
+        return None
+
+    [binomial_rv] = binomial_rvs
+
+    binomial_model_var = get_model_var_of_rv(fgraph, binomial_rv)
+    if binomial_model_var is None:
+        return None
+
+    # We want to replace free_rv by ConjugateRV()->(free_rv, conjugate_posterior_rv)
+    a, b = get_dist_params(beta_rv)
+    n, _ = get_dist_params(binomial_rv)
+
+    # Use value of y in new graph to avoid circularity
+    y = binomial_model_var.owner.inputs[1]
+
+    conjugate_a = a + y
+    conjugate_b = b + (n - y)
+    extra_dims = range(binomial_rv.type.ndim - beta_rv.type.ndim)
+    if extra_dims:
+        conjugate_a = conjugate_a.sum(extra_dims)
+        conjugate_b = conjugate_b.sum(extra_dims)
+    conjugate_beta_rv = Beta.dist(conjugate_a, conjugate_b)
```

Comment:
Would summing through data to make summary stats used be more general? 

For instance, y.sum(...) and (n - y).sum(...)

Reply:
That would mix independent dimensions of the beta. The model would no longer be equivalent to the original one

Reply:
Specifically a could itself not be a scalar

Reply:
I am saying that the parameter, p, and its shape defines the model and to keep that. Make n and y broadcastable with p only as that will make broadcastable with a and b. For instance,

a: ()
b: ()
p: ()
n: (100, )
y: (100, )

Then to sum n and y to match dim of p, (). n.sum(axis=0), y.sum(axis=0)

Another example :

a: (3, )
b: ()
p: (4, 3) # Defined from shape 
n: ()
y: (100, 4, 1)

n summed to something broadcastable with (4, 3)
n, y.sum(axis=0)

Reply:
```python
# Determined shape from p
y_stat = ...
n_stat = ...

# Cookie-cutter conjugate logic
a_post = a + y_stat
b_post = b + (n_stat - y_stat)
# Translate the conjugate parameters to PyMC distribution
conjugate_beta_rv = Beta.dist(alpha=a_post, beta=b_post) # Same shape as p before. Maybe shape needs to be passed as well

Reply:
Yup, that was the goal originally but as you found out it was not implemented correctly. I was only handing new dims. The last push has the `sum_bcast_dims` that reduces a_post/b_post to the shape of the original beta RV

Reply:
> a: (3, )
> b: ()
> p: (4, 3) # Defined from shape
> n: ()
> y: (100, 4, 1)

This example is not possible/not supported. the (100, 4, 1) cannot map to the (4, 3) of the p, unless you slice or something, which we don't allow

Reply:
Sure, but in the operation to get to posterior parameters, they are fine. 

Also, why not pass shape to new Beta.dist?

Reply:
> Sure, but in the operation to get to posterior parameters, they are fine and broadcast to the desired shape.

It matters how that shape went from 3 to 1. Where they summed/sliced/multiplied. Was there another operation in between like exponentiation/ reciprocal?

By only allowing direct use or left expand dims I rule out all these cases that I wouldn't know how to conjugate.

> Also, why not pass shape to new Beta.dist?

Yup, it's doing that now.